### ISSUE TITLE
async api on Datastore trait?

### ISSUE THREAD

**Issue by jasongoodwin** [indradb/indradb#234]
Hey, I was trying to do some work concurrently inside a `Datastore` implementation and noticed it's rather tricky as you can't block on the thread without risking a deadlock in the tokio threadpool.
Because the work will always be done in the context of a tokio runtime (eg via the server and its stack), the sync api on the Datastore makes it pretty tricky to actually do concurrent work, block on some tasks, and return the result without very carful handling.
Just curious what your thoughts are on making the `Datastore` trait async?

I gave it a quick shot to see how feasible it is - it's not a massive change. Plugins and Bench are missing and test-suite doesn't quite run yet, but i've actually run an implementation like this and tested it a bit running against 9 shards and doing operations concurrently in the lifecycle of a request.

I'm assuming the api was implemented as sync because async traits don't really exist in rust, and all of the operations in eg rocks and in memory store are sync. Once you try to do some concurrent work it becomes pretty clear it's not easy inside the context of the tokio runtime as you'll lock the threadpool easily anytime you try to block from the context. Using the tokio handle's block or trying to create other runtimes inside the tokio runtime cause panics from tokio etc.

I'm also trying to figure out if `spawn_blocking` from the server would be enough to prevent deadlocking but have to try it out as it'll still be trying to block from within the runtime, but the work won't be in the main tokio threadpool.

Let me know if you have any thoughts or insights on this item.

Couple related threads where people have encountered similar headache:
https://github.com/tokio-rs/tokio/issues/2376
https://github.com/seanmonstar/reqwest/issues/1215

**Comment by ysimonson** [comment]
I don't think making the core async makes sense. As you pointed out, IndraDB strictly calls into non-async under the hood. Making the trait async would just introduce unnecessary overhead. While it would be more convenient for async end-users, it would make it entirely unusable for non-async end-users.

FWIW the server has to deal with this same issue. It just calls into [`tokio::spawn`](https://github.com/indradb/indradb/blob/f6067ce59000efe317d99eed5a20a7cd2e771789/proto/src/server.rs#L235-L237), though `tokio::spawn_blocking` is better if you can use it (server doesn't because it needs to send responses on async channels.)

**Comment by jasongoodwin** [comment]
The problem with this is that _any_consumer of the api is unable to block the thread because it's an async -> sync call in a tokio runtime. You simply cannot do anything async in this context without it deadlocking threadpools or making tokio panic. So the server itself can control this a bit as it's calling into the async calls, but to anyone providing an api, I can't see an easy way to do this. 

For a sync end user of the api, simply slapping async on the method should be enough. 
See for example: 
https://github.com/jasongoodwin/indradb/pull/1/files#diff-c1a5c140b2388a4e3301d828a8d81dcb399d7866511593e7fd326962d490297f
or
https://github.com/jasongoodwin/indradb/pull/1/files#diff-a1d792f5ec483822fed9333082798aee80aef6ac51f3be05c412a26529539e80

For a consumer of the api, they don't control the code. That's what I'm thinking.
The alternative would be to make the calls out of server to call `spawn_blocking` but as a consumer of the api, I don't have that control so any mechanism I might hope to block the tokio thread will either result in panic or deadlock.

Maybe there is another way I'm not seeing though!

**Comment by ysimonson** [comment]
> For an end-user of the api, simply slapping async on the method should be enough.

Marking a function as async tells the end-user that (1) the function is async-friendly, i.e. will yield where appropriate, and (2) the function returns a future. Neither of these are true for IndraDB with the current implementations. So this is "lying" to async end-users, and adding unnecessary overhead for non-async end-users.

> The best solution I can find is just to do everything sync and avoid async at the cost of slightly longer response times, and honestly that seems fair too until you're doing large hunks of querying that could be concurrent.

Where possible, this is great. Are you sure that a sync setup actually performs worse? Given IndraDB's current datastore implementations "need" a dedicated thread, it's very likely that this actually performs better (assuming you employ, e.g. thread pools and other techniques to bring down the cost of dedicated thread), since async carries its own overhead. The only reason that the IndraDB server itself doesn't do this is because the underlying library for gRPC is async.

Failing that, you can still use IndraDB from an async runtime. Just use `spawn_blocking` where possible, and enable tokio's multithreaded runtime.

**Comment by jasongoodwin** [comment]
Yeah gotcha - thanks for taking the time to give some more insight here. 
I'll try to work around the deadlocking issue - I can't spawn_blocking from inside the async context and block the thread, so I'll probably send a small pr for the indra server. I'll try it out in a few.

Thanks again.

**Comment by jasongoodwin** [comment]
Was able to get beyond the deadlock with that pr and prove out a pretty decent 25% perf improvement with the concurrency across shards in a write heavy load. I'll gain some more confidence and clean up the pr if you're interested in accepting it.
